### PR TITLE
Mistake IOChaos

### PR DESCRIPTION
**Pull request by AsterNighT**
### What is changed and how does it work?
A new type of ioChaos is introduced. It generate random mistakes in read/write calls.
### Checklist

Tests

- [ ] Unit test
- [x] E2E test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Related changes

- [ ] Need to update the documentation

### Does this PR introduce a user-facing change?

```release-note
Introduce a new type of IOChaos: Mistake, which generates random error in read/write data
```

### CODE REVIEW

**Comment by Colstuwjx** [comment]
Just a reminder: don't forget to update the related crd yamls accordingly, see [Develop a New Chaos #Register the CRD](https://chaos-mesh.org/docs/development_guides/develop_a_new_chaos#register-the-crd) :)

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+	"k8s.io/klog"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/test/e2e/util"
+)
+
+func TestcaseIOMistakeDurationForATimeThenRecover(
```
Comment (YangKeao) [inline test/e2e/chaos/iochaos/io_mistake.go:33]: They should be added to `/test/e2e/chaos/basic.go`. These functions will not run automatically :crying_cat_face: .

Comment (AsterNighT) [reply test/e2e/chaos/iochaos/io_mistake.go:33]: got it

**Comment by g1eny0ung** [comment]
There are several debug prints (`fmt.Println`) in the test files. I'm not sure whether they need to be removed before the PR merge.

File: api/v1alpha1/podiochaos_types.go [hunk 2]
```diff
@@ -124,6 +131,24 @@ type AttrOverrideSpec struct {
...
 }
 
+// MistakeSpec represents one type of mistake
+type MistakeSpec struct {
+	//+optional
+	Filling FillingType `json:"filling,omitempty"`
+	//+optional
+	MaxOccurrences uint64 `json:"maxOccurrences,omitempty"`
+	//+optional
+	MaxLength uint64 `json:"maxLength,omitempty"`
```
[the remark this answers is not in the document]
Comment (AsterNighT) [reply api/v1alpha1/podiochaos_types.go]: sure thing

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+		if err == nil && res {
+			return true, nil
+		}
+		return false, nil
+	})
+	framework.ExpectNoError(err, "io chaos doesn't work as expected")
+
+	// cleanup
+	cli.Delete(ctx, ioChaos)
+	cancel()
```
[the remark this answers is not in the document]
Comment (AsterNighT) [reply test/e2e/chaos/iochaos/io_mistake.go]: indeed. This part is directly copy-pasted from io_error.go so I suppose same issue there too.

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+			Selector: v1alpha1.SelectorSpec{
+				Namespaces:     []string{ns},
+				LabelSelectors: map[string]string{"app": "io"},
+			},
+			Action:     v1alpha1.IoMistake,
+			Mode:       v1alpha1.OnePodMode,
+			VolumePath: "/var/run/data",
+			Path:       "/var/run/data/*",
+			Percent:    100,
+			// errno 5 is EIO -> I/O error
```
[the remark this answers is not in the document]
Comment (AsterNighT) [reply test/e2e/chaos/iochaos/io_mistake.go]: deleted, this is irrelevent

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
```
[the remark this answers is not in the document]
Comment (AsterNighT) [reply test/e2e/chaos/iochaos/io_mistake.go]: fixed

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
```
[the remark this answers is not in the document]
Comment (AsterNighT) [reply test/e2e/chaos/iochaos/io_mistake.go]: fixed

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+			return false, nil
+		}
+		if err == nil && !res {
+			return true, nil
+		}
+		return false, nil
+	})
+	framework.ExpectNoError(err, "fail to recover io chaos")
+
+	cancel()
```
[the remark this answers is not in the document]
Comment (AsterNighT) [reply test/e2e/chaos/iochaos/io_mistake.go]: fixed

File: api/v1alpha1/podiochaos_types.go [hunk 8]
```diff
@@ -124,6 +131,32 @@ type AttrOverrideSpec struct {
 	Rdev *uint32 `json:"rdev,omitempty"`
 }
 
+// MistakeSpec represents one type of mistake
+type MistakeSpec struct {
+	// Filling determines what is filled in the miskate data.
+	//+optional
```
Comment (STRRL) [inline api/v1alpha1/podiochaos_types.go:137]:
```suggestion
	// +optional
```

Appending one whitespace at the beginning of the comment. 

Although it does not affect the code generation, we should keep the same style.

Comment (STRRL) [reply api/v1alpha1/podiochaos_types.go:137]: Similar situations in other fields.

File: api/v1alpha1/podiochaos_types.go [hunk 9]
```diff
@@ -124,6 +131,32 @@ type AttrOverrideSpec struct {
...
+	Filling FillingType `json:"filling,omitempty"`
+
+	// There will be [1, MaxOccurrences] segments of wrong data. If MaxOccurrences is 0, no mistake will be injected
+	//+optional
+	MaxOccurrences uint64 `json:"maxOccurrences,omitempty"`
+
+	// Max length of each wrong data segment
```
Comment (STRRL) [inline api/v1alpha1/podiochaos_types.go:144]:
```suggestion
	// Max length in bytes of each wrong data segment
```

**Review by STRRL — CHANGES_REQUESTED** [review]
rest LGTM

**Comment by AsterNighT** [comment]
PTAL @STRRL

**Review by STRRL — DISMISSED** [review]
LGTM

File: api/v1alpha1/iochaos_webhook.go [hunk 10]
```diff
@@ -126,3 +127,21 @@ func (in *IoChaosSpec) validatePercent(percentField *field.Path) field.ErrorList
...
+	mistake := in.Mistake
+	if in.Mistake == nil {
+		return allErrs
+	}
+	allErrs = append(allErrs, in.validatePercent(mistakeField.Child("percent"))...)
+	// Useless since they're unsigned values. But keep it here for further tweaks.
+	if mistake.MaxLength < 0 {
+		allErrs = append(allErrs, field.Invalid(mistakeField.Child("maxLength"), mistake.MaxLength, "maxLength should be no less than 0"))
+	}
+	if mistake.MaxOccurrences < 0 {
```
Comment (WangXiangUSTC) [inline api/v1alpha1/iochaos_webhook.go:142]: `MaxLength` and `MaxOccurrences` should > 0 when action is mistake

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+	})
+	framework.ExpectNoError(err, "io chaos doesn't work as expected")
+
+	err = cli.Delete(ctx, ioChaos)
+	framework.ExpectNoError(err, "failed to delete io chaos")
+
+	klog.Infof("success to perform io chaos")
+	err = wait.PollImmediate(5*time.Second, 1*time.Minute, func() (bool, error) {
+		res, err := getPodIoMistake(c, port)
+		// input/output error is errno 5
```
Comment (WangXiangUSTC) [inline test/e2e/chaos/iochaos/io_mistake.go:94]: should this comment be deleted?

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+		},
+	}
+	err = cli.Create(ctx, ioChaos)
+	framework.ExpectNoError(err, "create io chaos error")
+
+	klog.Info("create iochaos successfully")
+
+	err = wait.PollImmediate(5*time.Second, 1*time.Minute, func() (bool, error) {
+		res, err := getPodIoMistake(c, port)
+		// input/output error is errno 5
```
Comment (WangXiangUSTC) [inline test/e2e/chaos/iochaos/io_mistake.go:152]: ditto

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+			return true, nil
+		}
+		return false, err
+	})
+	framework.ExpectNoError(err, "check paused chaos failed")
+
+	// wait 1 min to check whether io delay still exists
+	err = wait.PollImmediate(5*time.Second, 1*time.Minute, func() (bool, error) {
+		res, err := getPodIoMistake(c, port)
+		// input/output error is errno 5
```
Comment (WangXiangUSTC) [inline test/e2e/chaos/iochaos/io_mistake.go:188]: ditto

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+		if chaos.Status.Experiment.Phase == v1alpha1.ExperimentPhaseRunning {
+			return true, nil
+		}
+		return false, err
+	})
+	framework.ExpectNoError(err, "check resumed chaos failed")
+
+	err = wait.PollImmediate(5*time.Second, 1*time.Minute, func() (bool, error) {
+		res, err := getPodIoMistake(c, port)
+		// input/output error is errno 5
```
Comment (WangXiangUSTC) [inline test/e2e/chaos/iochaos/io_mistake.go:216]: ditto

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+				Cron: "@every 10m",
+			},
+		},
+	}
+	err = cli.Create(ctx, ioChaos)
+	framework.ExpectNoError(err, "create io chaos")
+
+	err = wait.PollImmediate(5*time.Second, 1*time.Minute, func() (bool, error) {
+		res, err := getPodIoMistake(c, port)
+		// input/output error is errno 5
```
Comment (WangXiangUSTC) [inline test/e2e/chaos/iochaos/io_mistake.go:276]: ditto

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+	})
+	framework.ExpectNoError(err, "io chaos doesn't work as expected")
+
+	err = cli.Delete(ctx, ioChaos)
+	framework.ExpectNoError(err, "failed to delete io chaos")
+
+	klog.Infof("success to perform io chaos")
+	err = wait.PollImmediate(5*time.Second, 1*time.Minute, func() (bool, error) {
+		res, err := getPodIoMistake(c, port)
+		// input/output error is errno 5
```
Comment (WangXiangUSTC) [inline test/e2e/chaos/iochaos/io_mistake.go:293]: ditto

File: test/e2e/chaos/iochaos/io_mistake.go [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+				Cron: "@every 10m",
+			},
+		},
+	}
+	err = cli.Create(ctx, ioChaos)
+	framework.ExpectNoError(err, "create io chaos")
+
+	err = wait.PollImmediate(5*time.Second, 1*time.Minute, func() (bool, error) {
+		res, err := getPodIoMistake(c, port)
+		// input/output error is errno 5
```
Comment (WangXiangUSTC) [inline test/e2e/chaos/iochaos/io_mistake.go:77]: ditto

**Review by WangXiangUSTC — COMMENTED** [review]
rest LGTM

File: api/v1alpha1/podiochaos_types.go [hunk 18]
```diff
@@ -124,6 +131,32 @@ type AttrOverrideSpec struct {
...
+	Filling FillingType `json:"filling,omitempty"`
+
+	// There will be [1, MaxOccurrences] segments of wrong data. If MaxOccurrences is 0, no mistake will be injected
+	// +optional
+	MaxOccurrences uint64 `json:"maxOccurrences,omitempty"`
+
+	// Max length of each wrong data segment in bytes
+	// +optional
+	MaxLength uint64 `json:"maxLength,omitempty"`
+}
```
Comment (STRRL) [inline api/v1alpha1/podiochaos_types.go:147]: I just realize that kubernetes do not recommend use unsigned integer:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types.

PTAL @AsterNighT ;

Comment (AsterNighT) [reply api/v1alpha1/podiochaos_types.go:147]: Oops, fresh knowledge to me, thanks!

Comment (AsterNighT) [reply api/v1alpha1/podiochaos_types.go:147]: Oh that'll lead to a disaster. There're tons of uint16/32/64 in out code. What're we gonna do about that?

Comment (STRRL) [reply api/v1alpha1/podiochaos_types.go]: I think just keep not using unsigned int in `api/v1alpha1`, and use webhook instead.

I try to count that, it is about 15- 20 refs to `uint` things, it's not so big.  🙈

Comment (AsterNighT) [reply api/v1alpha1/podiochaos_types.go]: I've fixed my uints, but I'm not sure what it is for the others. 🤦

**Review by WangXiangUSTC — DISMISSED** [review]
LGTM

**Comment by AsterNighT** [comment]
@STRRL @YangKeao PTAL

File: api/v1alpha1/iochaos_webhook.go [hunk 19]
```diff
@@ -126,3 +127,21 @@ func (in *IoChaosSpec) validatePercent(percentField *field.Path) field.ErrorList
...
+		return allErrs
+	}
+	allErrs = append(allErrs, in.validatePercent(mistakeField.Child("percent"))...)
+	// Useless since they're unsigned values. But keep it here for further tweaks.
+	if mistake.MaxLength <= 0 {
+		allErrs = append(allErrs, field.Invalid(mistakeField.Child("maxLength"), mistake.MaxLength, "maxLength should be greater than 0"))
+	}
+	if mistake.MaxOccurrences <= 0 {
+		allErrs = append(allErrs, field.Invalid(mistakeField.Child("maxOccurrences"), mistake.MaxOccurrences, "maxOccurrences should be greater than 0"))
+	}
```
Comment (STRRL) [inline api/v1alpha1/iochaos_webhook.go:144]: I prefer to put validation in the definition of that CRD with `// +kubebuilder:validation:<something>`.

Reference: https://book.kubebuilder.io/reference/markers/crd-validation.html

**Review by STRRL — CHANGES_REQUESTED** [review]
hi @AsterNighT, PTAL

File: api/v1alpha1/iochaos_webhook.go (hunk 19 above)
Comment (STRRL) [reply api/v1alpha1/iochaos_webhook.go:144]: We have used these markers in many places, you could find them by searching `+kubebuilder:validation`. :)

Comment (AsterNighT) [reply api/v1alpha1/iochaos_webhook.go:144]: Cool!

Comment (STRRL) [reply api/v1alpha1/iochaos_webhook.go]: Yes. 😁

File: api/v1alpha1/podiochaos_types.go [hunk 20]
```diff
@@ -124,6 +131,35 @@ type AttrOverrideSpec struct {
...
+type MistakeSpec struct {
+	// Filling determines what is filled in the miskate data.
+	// +optional
+	// +kubebuilder:validation:Enum=zero;random
+	Filling FillingType `json:"filling,omitempty"`
+
+	// There will be [1, MaxOccurrences] segments of wrong data. If MaxOccurrences is 0, no mistake will be injected
+	// +optional
+	// +kubebuilder:validation:Minimum=1
+	MaxOccurrences int64 `json:"maxOccurrences,omitempty"`
```
Comment (STRRL) [inline api/v1alpha1/podiochaos_types.go:144]: If we set `Mininum=1`, how about remove the comments about `MaxOccurrences = 0`

Comment (AsterNighT) [reply api/v1alpha1/podiochaos_types.go]: fixed

**Review by STRRL — APPROVED** [review]
LGTM

**Review by WangXiangUSTC — APPROVED** [review]
LGTM

**Pull request merged**